### PR TITLE
ts/modules-annotations

### PR DESCRIPTION
**Pull request by bre1470**
- Migrating annotations to TypeScript.

### CODE REVIEW

File: ts/annotations/MockPoint.ts [hunk 1]
```diff
@@ -0,0 +1,541 @@
...
+ */
+declare global {
+    namespace Highcharts {
+        interface AnnotationMockLabelOptionsObject {
+            x: (number|null);
+            y: (number|null);
+            point: AnnotationMockPoint;
+        }
+        class AnnotationMockPoint {
+            public static fromPoint(point: AnnotationPoint): AnnotationMockPoint;
```
Comment (pawelfus) [inline ts/annotations/MockPoint.ts]: `point: AnnotationPoint`: shouldn't be `point: Highcharts.Point` ? This static methods creates a new MockPoint from `series.points[index]` point.

File: ts/annotations/MockPoint.ts [hunk 2]
```diff
@@ -0,0 +1,541 @@
...
+        }
+        interface AnnotationMockSeries {
+            chart: AnnotationChart;
+            getPlotBox: Series['getPlotBox'];
+            xAxis?: (Axis|null);
+            yAxis?: (Axis|null);
+            visible: boolean;
+        }
+        interface AnnotationPoint {
+            mock?: undefined;
```
Comment (pawelfus) [inline ts/annotations/MockPoint.ts]: Actually `AnnotationPoint` is just `Highcharts.Point`. The way to distinguish between `Point` and `MockPoint` is that `MockPoint.mock` is defined and set to true. Should we then replace `AnnotationPoint` with `Point`?

File: ts/annotations/annotations.src.ts [hunk 3]
```diff
@@ -0,0 +1,1608 @@
...
+            symbol: string;
+            visible: boolean;
+            width: number;
+        }
+        interface AnnotationControlPointPositioner {
+            (target: AnnotationControllable): PositionObject;
+        }
+        type AnnotationDraggableValue = (''|'x'|'y'|'xy');
+        type AnnotationLabelType = AnnotationControllableLabel;
+        type AnnotationShapeType = AnnotationControllable;
```
Comment (pawelfus) [inline ts/annotations/annotations.src.ts]: I know it's not finished, just a heads-up: `AnnotationShapeType` can be probably one of many (`ControllableRect`, `ControllablePath`, `ControllableCircle` etc.)

File: ts/annotations/annotations.src.ts [hunk 4]
```diff
@@ -0,0 +1,1608 @@
...
+        interface AnnotationMockPointOptionsObject {
+            x: number;
+            y: number;
+            xAxis?: (number|string|Axis|null);
+            yAxis?: (number|string|Axis|null);
+        }
+        interface AnnotationPoint extends Point {
+            series: AnnotationSeries;
```
Comment (pawelfus) [inline ts/annotations/annotations.src.ts]: Ah, here is the definition for `AnnotationPoint`. One of other comments (about `mock`) doesn't matter, sorry.

File: ts/annotations/annotations.src.ts [hunk 5]
```diff
@@ -0,0 +1,1608 @@
...
+            text?: string;
+            useHTML: boolean;
+            verticalAlign: VerticalAlignValue;
+            x: number;
+            y: number;
+        }
+        interface AnnotationsLabelsOptions extends AnnotationsLabelOptions {
+            point?: (string|AnnotationMockPointOptionsObject);
+        }
+        interface AnnotationsOptions extends AnnotationControllableOptionsObject {
```
Comment (pawelfus) [inline ts/annotations/annotations.src.ts]: I'm not sure about this one. Are really `AnnotationsOptions` extended by `AnnotationControllableOptionsObject`? I trust you here, so could you double check this?

File: ts/annotations/annotations.src.ts [hunk 6]
```diff
@@ -0,0 +1,1608 @@
...
+            controlPointOptions: AnnotationControlPointOptionsObject;
+            draggable: AnnotationDraggableValue;
+            events: AnnotationsEventsOptions;
+            id?: (number|string);
+            labelOptions: AnnotationsLabelOptions;
+            labels: Array<AnnotationsLabelsOptions>;
+            shapeOptions: AnnotationsShapeOptions;
+            shapes: Array<AnnotationsShapesOptions>;
+            typeOptions: AnnotationsTypeOptions;
```
Comment (pawelfus) [inline ts/annotations/annotations.src.ts]: I think these `typeOptions` should be defined in specific files (e.g. Fibonacci, ElliottWave) etc. Every type can provide different set of options.

File: ts/annotations/controllable/ControllableLabel.ts [hunk 7]
```diff
@@ -0,0 +1,561 @@
...
+ */
+declare global
+{
+    namespace Highcharts
+    {
+        interface AnnotationAlignObject extends AlignObject {
+            height?: number;
+            width?: number;
+        }
+        class AnnotationControllableLabel implements AnnotationControllable {
```
Comment (pawelfus) [inline ts/annotations/controllable/ControllableLabel.ts]: Wouldn't be better to use `extends` ? There will be more classes that can use `AnnotationControllable` base and overwrite it's methods.

File: ts/annotations/eventEmitterMixin.ts [hunk 8]
```diff
@@ -0,0 +1,349 @@
...
+ * */
+
+import H from '../parts/Globals.js';
+
+/**
+ * Internal types.
+ */
+declare global {
+    namespace Highcharts {
+        interface AnnotationEventEmitter extends AnnotationEventEmitterMixin {
```
[the remark this answers is not in the document]
Comment (bre1470) [reply ts/annotations/eventEmitterMixin.ts]: The mixin describes only the export type, while the emitter describes the instance for mixin functions. Not everything of the emitter is part of the mixin.

File: ts/annotations/MockPoint.ts (hunk 1 above)
Comment (bre1470) [reply ts/annotations/MockPoint.ts]: `AnnotationPoint` is only an internal interface to describe the `Point` instance during annotation processing, so that `any` casts can be avoided. Same for `AnnotationChart` and `AnnotationSeries`.

File: ts/annotations/MockPoint.ts (hunk 2 above)
Comment (bre1470) [reply ts/annotations/MockPoint.ts]: See previous comment.

File: ts/annotations/annotations.src.ts (hunk 3 above)
Comment (bre1470) [reply ts/annotations/annotations.src.ts]: Thanks! 👍

File: ts/annotations/annotations.src.ts (hunk 4 above)
Comment (bre1470) [reply ts/annotations/annotations.src.ts]: See my second comment.

File: ts/annotations/annotations.src.ts (hunk 5 above)
Comment (bre1470) [reply ts/annotations/annotations.src.ts]: Well yesterday it was the conclusion, as all controllables take AnnotationsOptions, but I really do not know. It is just the typing that leads to this conclusion.

File: ts/annotations/controllable/ControllableLabel.ts (hunk 7 above)
Comment (bre1470) [reply ts/annotations/controllable/ControllableLabel.ts]: AnnotationControllable is not a class - just a mixin object - you can not extend a class from an interface.

File: ts/annotations/eventEmitterMixin.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (pawelfus) [reply ts/annotations/eventEmitterMixin.ts]: I see, thanks!

File: ts/annotations/annotations.src.ts (hunk 5 above)
Comment (pawelfus) [reply ts/annotations/annotations.src.ts]: Ah, right, this `x`/`y` options that actually works. Thanks!

File: ts/annotations/controllable/ControllableLabel.ts (hunk 7 above)
Comment (pawelfus) [reply ts/annotations/controllable/ControllableLabel.ts]: Ah, okay!

File: ts/annotations/MockPoint.ts (hunk 1 above)
Comment (pawelfus) [reply ts/annotations/MockPoint.ts]: Yes, realized that after writing this comment, thanks!

File: ts/annotations/annotations.src.ts (hunk 6 above)
Comment (bre1470) [reply ts/annotations/annotations.src.ts]: As you can merge interfaces but not property typing, I kept the generic type for inheritance purpose.

File: ts/annotations/navigationBindings.ts [hunk 9]
```diff
@@ -0,0 +1,1626 @@
...
+            public initEvents(): void;
+            public initUpdate(): void;
+            public removeEvents(): void;
+            public update(options: NavigationOptions): void
+        }
+        interface NavigationBindingsButtonEventsObject {
+            button: HTMLDOMElement;
+            events: NavigationBindingsOptionsObject;
+        }
+        interface NavigationBindingsOptionsObject {
```
Comment (pawelfus) [inline ts/annotations/navigationBindings.ts]: Not available in API: http://localhost:9005/class-reference/Highcharts.NavigationBindingsOptionsObject

Referenced for example from: http://localhost:9005/highstock/navigation.bindings.arrowInfinityLine

File: ts/modules/marker-clusters.src.ts [hunk 10]
```diff
@@ -288,6 +288,8 @@ var Series = H.Series,
  * @product      highcharts highmaps
  * @since        next
  * @optionparent plotOptions.scatter.cluster
+ *
+ * @private
```
Comment (pawelfus) [inline ts/modules/marker-clusters.src.ts]: Is this related to Annotations?

File: ts/annotations/controllable/markerMixin.ts [hunk 11]
```diff
@@ -0,0 +1,227 @@
...
+        interface AnnotationMarkerMixin {
+            markerEndSetter(this: SVGElement, value: string): void;
+            markerStartSetter(this: SVGElement, value: string): void;
+            setItemMarkers(this: AnnotationControllablePath, item: AnnotationControllablePath): void;
+        }
+        interface AnnotationChart {
+            afterGetContainer(): void;
+        }
+        interface Options {
+            defs?: Dictionary<SVGDefinitionObject>;
```
Comment (pawelfus) [inline ts/annotations/controllable/markerMixin.ts]: Aren't `defs` already defined? See: https://api.highcharts.com/highcharts/defs

File: ts/annotations/navigationBindings.ts (hunk 9 above)
Comment (bre1470) [reply ts/annotations/navigationBindings.ts]: I do not understand. Should this internal interface be public or should this not be created internally?

File: ts/modules/marker-clusters.src.ts (hunk 10 above)
Comment (bre1470) [reply ts/modules/marker-clusters.src.ts]: This is related to the class references. The tag gets ignored by the API scripts and just makes sure, that JSDoc is not picking it up itself (to put it in the global namespace).

File: ts/annotations/controllable/markerMixin.ts (hunk 11 above)
Comment (bre1470) [reply ts/annotations/controllable/markerMixin.ts]: Yes, but this is about the internal namespace, not about the API documentation.

File: ts/annotations/navigationBindings.ts (hunk 9 above)
Comment (pawelfus) [reply ts/annotations/navigationBindings.ts]: I think this should be public. Since here: http://localhost:9005/class-reference/Highcharts.NavigationBindingsOptionsObject - we can click reference to this interface:

Screen Shot 2020-01-06 at 11 09 58

I'm pretty sure we users don't want to see this after clicking:
Screen Shot 2020-01-06 at 11 11 57

Comment (bre1470) [reply ts/annotations/navigationBindings.ts]: It should be fixed now.
